protocol/auditlog.go

```diff
@@ -0,0 +1,187 @@
+func (l *ConiksAuditLog) IsKnownDirectory(addr string) bool {
+	h := l.histories[addr]
+	if h != nil {
+		return true
+	}
+	return false
+}
+
+// Insert creates a new directory history for the key directory addr
+// and inserts it into the audit log l.
+// The directory history is initialized with the key directory's
+// signing key signKey, a list of STRs representing the
+// directory's prior history oldSTRs, and the directory's latest STR
+// latestSTR.
+// Insert() returns an ErrAuditLog if the auditor attempts to create
+// a new history for a known directory, an ErrMalformedDirectoryMessage
+// if oldSTRs is malformed, and nil otherwise.
+// Insert() only creates the initial entry in the log for addr. Use Update()
+// to insert newly observed STRs for addr in subsequent epochs.
+// FIXME: pass Response message as param
+// masomel: will probably want to write a more generic function
+// for "catching up" on a history in case an auditor misses epochs
+func (l *ConiksAuditLog) Insert(addr string, signKey sign.PublicKey,
+	oldSTRs map[uint64]*DirSTR, latestSTR *DirSTR) error {
+
+	// error if we want to create a new entry for an addr we already know
+	if l.IsKnownDirectory(addr) {
+		return ErrAuditLog
+	}
+
+	// make sure we have oldEpochs unless we're inserting right
+	// at the start of a directory's history
+	if oldSTRs == nil && latestSTR.Epoch > 0 {
+		return ErrMalformedDirectoryMessage
+	}
+
+	// create the new directory history
+	h := newDirectoryHistory(signKey, latestSTR)
+
+	startEp := uint64(0)
+	endEp := latestSTR.Epoch
+
+	// add each old STR into the history
+	// This loop automatically catches if oldSTRs is malformed
+	// (i.e. oldSTRs starts at epoch 0 and
+	// len(oldSTRs) != latestSTR.Epoch-1)
+	for ep := startEp; ep < endEp; ep++ {
+		str := oldSTRs[ep]
+		if str == nil {
+			return ErrMalformedDirectoryMessage
+		}
+		h.snapshots[ep] = str
+	}
+
+	// don't forget to add the latest STR
+	h.snapshots[endEp] = latestSTR
+
+	l.histories[addr] = h
+
+	// FIXME: verify the consistency of each new STR
+	return nil
+}
+
+// Update verifies the consistency of a newly observed STR newSTR for
+// the directory addr, and inserts the newSTR into addr's directory history
+// if the checks (i.e. STR signature and hash chain verifications) pass.
+// Update() returns nil if the checks pass, and the appropriate consistency
+// check error otherwise. Update() assumes that Insert() has been called for
+// addr prior to its first call and thereby expects that an entry for addr
+// exists in the audit log l.
+// FIXME: pass Response message as param
+func (l *ConiksAuditLog) Update(addr string, newSTR *DirSTR) error {
+
+	// error if we want to update the entry for an addr we don't know
+	if !l.IsKnownDirectory(addr) {
+		return ErrAuditLog
+	}
+
+	h := l.histories[addr]
+
+	if err := verifySTRConsistency(h.signKey, h.latestSTR, newSTR); err != nil {
+		return err
+	}
+
+	// update the latest STR
+	h.snapshots[h.latestSTR.Epoch] = h.latestSTR
```

Comment:
The question I'm asking is when should the newSTR be added to the map of snapshots.  At Update, or at the next call to Update.  Because right now it's only on subsequent calls.  Except for at Insert, when it gets added right away.